qiskit/transpiler/target.py

```diff
@@ -1048,11 +1051,22 @@ def build_coupling_map(self, two_q_gate=None):
         # existing and return
         if self._coupling_graph is not None:
             cmap = CouplingMap()
-            cmap.graph = self._coupling_graph
+            if filter_idle_qubits:
+                cmap.graph = self._filter_coupling_graph()
+            else:
+                cmap.graph = self._coupling_graph
             return cmap
         else:
             return None
 
+    def _filter_coupling_graph(self):
+        has_operations = set(itertools.chain.from_iterable(self.qargs))
+        graph = self._coupling_graph.copy()
+        for index in graph.node_indices():
+            if index not in has_operations:
+                graph.remove_node(index)
```

Comment:
This could save some work and allocation in case none need to be removed.
```suggestion
        graph = self._coupling_graph
        toremove = set(graph.node_indices()).difference(has_operations)
        if len(toremove) > 0:
            graph = graph.copy()
            graph.remove_nodes_from(list(toremove))
```

Reply:
Yeah, it's fine to share the graph object. By default the output of `build_coupling_map()` is cached to avoid reconstructing it multiple times when the instructions change. The reason I had to copy here was to avoid mutating the cached copy since we don't want to do that except for when we're explicitly handling the holes by calling `connected_components`